include/irr/asset/material_compiler/IR.h

```diff
@@ -0,0 +1,233 @@
+        //union {
+            STextureSource texture;
+        //};
+    };
+
+    struct CEmissionNode : INode
+    {
+        CEmissionNode() : INode(ES_EMISSION) {}
+
+        float intensity;
+    };
+
+    struct CBSDFCombinerNode : INode
+    {
+        enum E_TYPE
+        {
+            //mix of N BSDFs
+            ET_MIX,
+            //blend of 2 BSDFs weighted by constant or texture
+            ET_WEIGHT_BLEND,
+            //blend of 2 BSDFs weighted by fresnel
+            ET_FRESNEL_BLEND,
+            //blend of 2 BSDFs weighted by custom direction-based curve
+            ET_CUSTOM_CURVE_BLEND,
+            ET_DIFFUSE_AND_SPECULAR
+        };
+
+        E_TYPE type;
+
+        CBSDFCombinerNode(E_TYPE t) : INode(ES_BSDF_COMBINER), type(t) {}
+    };
+    struct CBSDFBlendNode : CBSDFCombinerNode
+    {
+        CBSDFBlendNode() : CBSDFCombinerNode(ET_WEIGHT_BLEND) {}
+
+        SParameter<float> weight;
+    };
+    struct CBSDFMixNode : CBSDFCombinerNode
+    {
+        CBSDFMixNode() : CBSDFCombinerNode(ET_MIX) {}
+
+        using weights_t = core::smart_refctd_dynamic_array<float>;
+        weights_t weights;
+    };
+
+    struct CBSDFNode : INode
+    {
+        enum E_TYPE
+        {
+            ET_DIFFTRANS,
+            ET_SPECULAR_DELTA,
+            ET_MICROFACET_DIFFUSE,
+            ET_MICROFACET_SPECULAR,
+            ET_SHEEN,
+            ET_COATING
+        };
+
+        CBSDFNode(E_TYPE t) : INode(ES_BSDF), type(t) {}
+
+        E_TYPE type;
+    };
+    struct CMicrofacetSpecularBSDFNode : CBSDFNode
+    {
+        enum E_NDF
+        {
+            ENDF_BECKMANN,
+            ENDF_GGX,
+            ENDF_ASHIKHMIN_SHIRLEY,
+            ENDF_PHONG
+        };
+        enum E_SHADOWING_TERM
+        {
+            EST_SMITH,
+            EST_VCAVITIES
+        };
+        enum E_SCATTER_MODE
+        {
+            ESM_REFLECT,
+            ESM_TRANSMIT
+        };
+
+        CMicrofacetSpecularBSDFNode(E_TYPE t = ET_MICROFACET_SPECULAR) : CBSDFNode(t) {}
+
+        E_NDF ndf;
+        E_SHADOWING_TERM shadowing;
+        E_SCATTER_MODE scatteringMode;
+        SParameter<float> alpha_u;
+        SParameter<float> alpha_v;
+    };
+    struct CMicrofacetDiffuseBSDFNode : CBSDFNode
+    {
+        CMicrofacetDiffuseBSDFNode() : CBSDFNode(ET_MICROFACET_DIFFUSE) {}
+
+        SParameter<color_t> reflectance;
+        SParameter<float> alpha_u;
+        SParameter<float> alpha_v;
+    };
+    struct CDifftransBSDFNode : CBSDFNode
+    {
+        CDifftransBSDFNode() : CBSDFNode(ET_DIFFTRANS) {}
+
+        SParameter<color_t> transmittance;
+    };
+    struct CCoatingBSDFNode : CMicrofacetSpecularBSDFNode
+    {
+        CCoatingBSDFNode() : CMicrofacetSpecularBSDFNode(ET_COATING) {}
+
+        SParameter<color_t> sigmaA;
+        float thickness;
+    };
+
+    core::smart_refctd_ptr<INode> root;
```

Comment:
its not a single tree, but a forest (we need this info for SIMD optimization in the backends) so you will have multiple roots

Reply:
hm? how is this a forest?

Reply:
so I guess w're clear now?

Reply:
yea